rfcs/017 - Cluster Data Storage RFC.md

```diff
@@ -0,0 +1,221 @@
+
+This can be broken into multiple tasks:
+
+* Creating the endpoints and the serialization/deserialization of the HandlerMemory necessary
+* Querying LogRendezvousHash for the first node to know which one is "in charge" of this data, or the first `n` nodes if it is itself so it knows where to push the mutation to
+* Fronting the datastore internal data structure with this logic instead of using it directly by the opcodes (when in daemon mode, leaving it as-is or having a "cluster" of one when running locally)
+* Exposing this new behavior into the docker container, and the library work necessary to make this feature usable and natural in other languages.
+* Adding atomic functionality to the datastore API and internals to make distributed mutation (not just key replacement) possible
+
+### Cluster Coordination Work
+
+The new endpoints have some extra logic that they need to tackle, as this is the first time the nodes of the cluster are actually communicating with each other. The LogRendezvousHash module needs to be updated on the *actual* state of the cluster versus the expected state defined in the DNS records: which nodes are actually up or down from its perspective? That determines which nodes are considered for ownership of the datastore data. RendezvousHash (and our own LogRendezvousHash) are stable, in that the top `n` nodes after one of them goes down will still contain `n - 1` of the original set, so not too much data thrashing occurs when a node goes down or comes online.
+
+Further, while a full network partition between two groups of nodes will result in a "split brain" scenario in an unavoidable fashion, with LogRendezvousHash it is possible to counteract a partial split by way of a request forwarding mechanism. If a node cannot directly reach another node, it should try to ask all other nodes in its list if any of them can reach the node. If so, the first one to respond in the positive becomes the new "router" that all requests to that node are forwarded through, with periodic attempts in the background to remove the extra hop by re-polling. Only if all nodes a node can reach report that they also cannot reach the node in question should it be marked as "down" and assignment of keys shuffled with the new, reduced set. This is fundamentally how the internet routes packets to unknown MAC addresses, and can guard against misconfigured firewalls/routers black hole-ing packets that they shouldn't be, but only works when there is an unaffected third party, so requires at least three separate sub-clusters to work.
```

Comment:
Seeing the data structure used to represent the "report" and "router" for a cluster that every node in it will have would be useful?

Reply:
I considered it an implementation detail? The nodes need to track metadata on all of the other nodes to determine what nodes are in charge of what and how to route between the nodes, the exact structure will depend on what is clean and efficient to write in Rust. This high level description felt like more than enough?

Reply:
Fair